// openvidu-browser-v2compatibility/lib/OpenVidu/EventDispatcher.js
"use strict";
/*
 * (C) Copyright 2017-2022 OpenVidu (https://openvidu.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventDispatcher = void 0;
const OpenViduLogger_1 = require("../OpenViduInternal/Logger/OpenViduLogger");
const EventEmitter = require("wolfy87-eventemitter");
/**
 * @hidden
 */
const logger = OpenViduLogger_1.OpenViduLogger.getInstance();
class EventDispatcher {
    constructor() {
        /**
         * @hidden
         */
        this.userHandlerArrowHandler = new WeakMap();
        /**
         * Emmits events to the user
         * @hidden
         */
        this.publicEmmiter = new EventEmitter();
    }
    /**
     * @hidden
     */
    onAux(type, message, handler) {
        const arrowHandler = (event) => {
            logger.debug(message, event);
            handler(event);
        };
        this.userHandlerArrowHandler.set(handler, arrowHandler);
        this.publicEmmiter.on(type, arrowHandler);
        return this;
    }
    /**
     * @hidden
     */
    onceAux(type, message, handler) {
        const arrowHandler = (event) => {
            logger.debug(message, event);
            handler(event);
            // Remove handler from map after first and only execution
            this.userHandlerArrowHandler.delete(handler);
        };
        this.userHandlerArrowHandler.set(handler, arrowHandler);
        this.publicEmmiter.once(type, arrowHandler);
        return this;
    }
    /**
     * @hidden
     */
    offAux(type, handler) {
        if (!handler) {
            this.publicEmmiter.removeAllListeners(type);
        }
        else {
            // Must remove internal arrow function handler paired with user handler
            const arrowHandler = this.userHandlerArrowHandler.get(handler);
            if (!!arrowHandler) {
                this.publicEmmiter.off(type, arrowHandler);
            }
            this.userHandlerArrowHandler.delete(handler);
        }
        return this;
    }
}
exports.EventDispatcher = EventDispatcher;
