// openvidu-browser-v2compatibility/lib/OpenViduInternal/Logger/OpenViduLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenViduLogger = void 0;
const ConsoleLogger_1 = require("./ConsoleLogger");
class OpenViduLogger {
    constructor() {
        this.defaultConsoleLogger = new ConsoleLogger_1.ConsoleLogger(globalThis.console);
        this.isProdMode = false;
    }
    /**
     * @hidden
     */
    static getInstance() {
        if (!OpenViduLogger.instance) {
            OpenViduLogger.instance = new OpenViduLogger();
        }
        return OpenViduLogger.instance;
    }
    /**
     * @hidden
     */
    log(...args) {
        if (!this.isProdMode) {
            this.defaultConsoleLogger.log.apply(this.defaultConsoleLogger.logger, arguments);
        }
    }
    /**
     * @hidden
     */
    debug(...args) {
        if (!this.isProdMode) {
            this.defaultConsoleLogger.debug.apply(this.defaultConsoleLogger.logger, arguments);
        }
    }
    /**
     * @hidden
     */
    info(...args) {
        if (!this.isProdMode) {
            this.defaultConsoleLogger.info.apply(this.defaultConsoleLogger.logger, arguments);
        }
    }
    /**
     * @hidden
     */
    warn(...args) {
        this.defaultConsoleLogger.warn.apply(this.defaultConsoleLogger.logger, arguments);
    }
    /**
     * @hidden
     */
    error(...args) {
        this.defaultConsoleLogger.error.apply(this.defaultConsoleLogger.logger, arguments);
    }
    enableProdMode() {
        this.isProdMode = true;
    }
}
exports.OpenViduLogger = OpenViduLogger;
