// openvidu-browser-v2compatibility/lib/OpenViduInternal/Interfaces/Private/InternalSignalTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InternalSignalType = void 0;
var InternalSignalType;
(function (InternalSignalType) {
    InternalSignalType["CUSTOM_CLIENT_SIGNAL_PREFIX"] = "custom-openvidu-client-signal:";
    InternalSignalType["CUSTOM_SERVER_SIGNAL_PREFIX"] = "custom-openvidu-server-signal:";
    InternalSignalType["OPENVIDU_STREAM_PROPERTY_CHANGED"] = "openvidu-stream-property-changed";
    InternalSignalType["OPENVIDU_CONNECTION_CREATED"] = "openvidu-connection-created";
    InternalSignalType["OPENVIDU_RECORDING_STARTED"] = "openvidu-recording-started";
    InternalSignalType["OPENVIDU_RECORDING_STOPPED"] = "openvidu-recording-stopped";
    InternalSignalType["OPENVIDU_BROADCASTING_STARTED"] = "openvidu-broadcasting-started";
    InternalSignalType["OPENVIDU_BROADCASTING_STOPPED"] = "openvidu-broadcasting-stopped";
    InternalSignalType["OPENVIDU_PARTICIPANT_PUBLISHED"] = "openvidu-participant-published";
    InternalSignalType["OPENVIDU_SIGNAL"] = "openvidu-internal-signal";
    InternalSignalType["OPENVIDU_FORCE_DISCONNECT"] = "openvidu-force-disconnect";
    InternalSignalType["OPENVIDU_FORCE_UNPUBLISH"] = "openvidu-force-unpublish";
    InternalSignalType["OPENVIDU_UNPUBLISH_STREAM"] = "openvidu-unpublish-stream";
})(InternalSignalType = exports.InternalSignalType || (exports.InternalSignalType = {}));
